### ISSUE TITLE
Seek operation is slow

### ISSUE THREAD

**Issue by alanhe** [colinmarc/hdfs#270]
Hi,

I have a big SequenceFile. I want to jump back and forth to the sync marks within it.
Comparing to the official HDFS Java client, I find the go implementation is much slower.
I guess it is because it always closes the existing connection and creates a new one.

https://github.com/colinmarc/hdfs/blob/f2f512db170db82ad41590c4ba3b7718b13317d2/file_reader.go#L158-L164

If I understand it right, the Java client buffers some data so that seek may be done without reconnection -- Do you think it's possible to make seek faster? Or, could you share me some tips of how to do it right?

**Comment by colinmarc** [comment]
Hi @alanhe,

How big are your seeks? Sorry for responding so much later!

**Comment by alanhe** [comment]
Thanks for replying!

I'm scanning through a SequenceFile. It is organized similar to a Parquet format. It has an embedded index appended to the tail of the file.
The file structure looks like this:
```
<timestamp1,   chunk1>
<timestamp2,   chunk2>
<timestamp3,   chunk3>
....
<timestampN-1, chunkN-1>
<timestampN,   chunkN>
<_,            index>
```
The maximum chunk size is 5MB, and the average value is around 1MB.

My scan process is split into to parts:

**Part1**
First, I jump to the end of the file parse and rebuild the index in memory.

**Part2**
Then, find the related chunk offsets and seek to the start of each chunks (position of a Sequence key position).

In **Part2**, the seek is always in timestamp order, either from latest-to-oldest, or from oldest-to-latest.
For example, if only chunk1, chunk3 and chunkN are related, then the seeking sequence is either one of,
1. `chunk1` -> `chunk3` -> `chunkN`, or
2. `chunkN` -> `chunk3` -> `chunk1`

I can understand why case#2 (seeking backward) is slower. Though I don't have a precise benchmark, my generally feeling is that, even for case#1, and other use cases, seeking by the go-implementation is slower than the java SDK.

**Comment by colinmarc** [comment]
Hi @alanhe,

Taking a look at the code, you're totally right - this is massively inefficient. To seek a few bytes forward, the FileReader throws out the block, reconnects to the datanode, and then discards data from the front of the block until we reach the offset again.

I'll push a fix for this, but in the meantime you can use `io.CopyN(io.Discard, f, off)` instead of `Seek`.

**Comment by colinmarc** [comment]
Fixed in 11bad6d.
